libflux/src/semantic/types.rs

```diff
@@ -58,13 +49,102 @@ impl fmt::Display for PolyType {
     }
 }
 
+// Kind constraints on type variables.
+//
+// Ex. forall [t0] where t0:Comparable + Equatable (x:t0, y:t0) -> bool
+//                       ^_______________________^
+//
+#[derive(Debug, Clone, PartialEq)]
+pub struct Const {
+    pub kinds: HashMap<Tvar, HashSet<Kind>>,
+}
+
+impl fmt::Display for Const {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        DisplayList {
+            values: self
+                .kinds
+                .iter()
+                .collect::<BTreeMap<_, _>>()
+                .iter()
+                .map(|(&tv, &kinds)| TvarConst { tv, kinds })
+                .collect(),
+            delim: ", ",
+        }
+        .fmt(f)
+    }
+}
+
+// Private struct whose sole purpose is for displaying the generic
+// constraints for the free variables of a polytype.
+#[derive(Debug)]
+struct TvarConst<'a> {
+    tv: &'a Tvar,
+    kinds: &'a HashSet<Kind>,
+}
+
+impl<'a> fmt::Display for TvarConst<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "{}:{}",
+            self.tv,
+            DisplayList {
+                values: self.kinds.iter().collect::<BTreeSet<_>>().iter().collect(),
+                delim: " + ",
+            },
+        )
+    }
+}
+
+// Unification equates two types by producing a substitution whereby
+// the types are equal after application of the substitution.
+//
+// More formally, s = unify(x,y) => s(x) = s(y) .
+//
+// Note here that unification also takes a set of constrained type
+// variables and produces a possibly updated set of constraints along
+// with the substitution equating the two types.
+//
+trait Unifiable {
+    fn unify(self, with: Self, cons: Const) -> Result;
+}
+
+// Types may be constrained with Kinds.
+//
+// Unlike unification which asserts an equality constraint on two
+// types, a kind constraint asserts the containment of a type within
+// a type class or family.
+//
+trait Constrainable {
+    fn constrain(self, with: Kind, cons: Const) -> Result;
+}
+
+type Result = result::Result<(Subst, Const), Error>;
+
+#[derive(Debug)]
+enum Error {
+    Constrain { t: MonoType, with: Kind },
+    Unify { t: MonoType, with: MonoType },
```

Comment:
Yes sorry this wasn't clear. I thought constrain was better verbiage for kinds since unify connotes equality more or less. I guess the doc comment for `Constrainable` didn't explain this very well?

I agree I have too many names with Constraint in them. I'll try to come up with more descriptive ones.